### PR TITLE
Adds improvements in organize imports. See added tests.

### PR DESCRIPTION
**Pull request by wpopielarski**
This PR adds some failing scenarios shown in newly added bulk of tests.

### CODE REVIEW

File: src/main/scala-2_11/scala/tools/refactoring/implementations/oimports/TreeToolboxScanners.scala [hunk 1]
```diff
@@ -29,5 +29,10 @@ class TreeToolboxScanners[G <: Global](val global: G) {
     }
 
     def comments = comments_.toList
+
+    override def emitIdentifierDeprecationWarnings = false
+
+    override def error(off: Offset, msg: String): Unit =
+      if (name != global.nme.MACROkw) super.error(off, msg)
```
Comment (sschaef) [inline src/main/scala-2_11/scala/tools/refactoring/implementations/oimports/TreeToolboxScanners.scala]: Why do you want to suppress errors from macros?

File: src/main/scala/scala/tools/refactoring/analysis/CompilationUnitDependencies.scala [hunk 2]
```diff
@@ -447,6 +457,23 @@ trait CompilationUnitDependencies extends CompilerApiExtensions with ScalaVersio
                   fakeSelectTreeFromType(tpe, sym, t.pos) match {
                     case s: Select if !isDefinedLocallyAndQualifiedWithEnclosingPackage(s) =>
                       addToResult(s)
+                    case i: Ident if newWay =>
+                      val realText = i.pos.source.content.slice(i.pos.start, i.pos.end).mkString
+                      val classOfRegex = """(?<=classOf\[)(.+)(?=\])""".r
```
Comment (sschaef) [inline src/main/scala/scala/tools/refactoring/analysis/CompilationUnitDependencies.scala]: Isn't this the same as `str.startsWith("classOf[") && str.endsWith("]")`? I think that is easier to read than the regex.

File: src/main/scala/scala/tools/refactoring/implementations/oimports/Region.scala [hunk 3]
```diff
@@ -70,8 +84,19 @@ object MiscTools {
   def isScalaLanguageImport[G <: Global](g: G): g.Import => Boolean = {
     import g._
     val language = newTermName("language")
+
+    @tailrec
+    def isScalaLanguage(qual: Select): Boolean = qual match {
```
Comment (sschaef) [inline src/main/scala/scala/tools/refactoring/implementations/oimports/Region.scala]: this method is duplicated. Can't you move it out to some common location?

File: src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala [hunk 4]
```diff
@@ -4410,4 +4411,547 @@ class OrganizeImportsTest extends OrganizeImportsBaseTest {
     class Tested(val id: Int) extends Extended[Int](id, "text")
     """ isNotModified
   } applyRefactoring organizeCustomized(dependencies = Dependencies.RecomputeAndModify)
+
+  @Test
+  def shouldNotRemoveImportIfInPathToInnerType() = new FileSet {
+    """
+    package acme
+
+	  trait Tupler[T]
```
Comment (sschaef) [inline src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala]: wrong indentation

File: src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala [hunk 5]
```diff
@@ -4410,4 +4411,547 @@ class OrganizeImportsTest extends OrganizeImportsBaseTest {
...
+    }
+    """ isNotModified
+  } applyRefactoring organizeCustomized(dependencies = Dependencies.RecomputeAndModify)
+
+  @Test
+  def shouldNotRemoveImportIfInPathToInheritedInnerType() = new FileSet {
+    """
+    package acme
+
+	  trait Tupler[T]
```
Comment (sschaef) [inline src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala]: wrong indentation

File: src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala [hunk 6]
```diff
@@ -4410,4 +4411,547 @@ class OrganizeImportsTest extends OrganizeImportsBaseTest {
...
+
+    class Tested {
+      import org.{acme => jlang}
+      val t = classOf[jlang.Integer]
+    }
+    """ isNotModified
+  } applyRefactoring organizeCustomized(dependencies = Dependencies.RecomputeAndModify)
+
+  @Test
+  def shouldNotRemoveImportWhenPackageRenamedAndUsedInClassOf_v5() = new FileSet {
```
Comment (sschaef) [inline src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala]: v5-v8 have nothing to do with `classOf` anymore

File: src/main/scala-2_11/scala/tools/refactoring/implementations/oimports/TreeToolboxScanners.scala (hunk 1 above)
Comment (wpopielarski) [inline src/main/scala-2_11/scala/tools/refactoring/implementations/oimports/TreeToolboxScanners.scala]: for this particular scanner only. It's role is just to pick comments so `macro` keyword should not hold it

File: src/main/scala/scala/tools/refactoring/analysis/CompilationUnitDependencies.scala (hunk 2 above)
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/analysis/CompilationUnitDependencies.scala]: regex looks nicer :) and extraction of text is cooler I guess

File: src/main/scala/scala/tools/refactoring/implementations/oimports/Region.scala (hunk 3 above)
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/implementations/oimports/Region.scala]: discussed, merge `CompilationUnitDependencies` with new approach should purge redundant implementation

File: src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala [hunk 7]
```diff
@@ -4410,4 +4411,902 @@ class OrganizeImportsTest extends OrganizeImportsBaseTest {
...
+  @Test
+  def shouldNotRemoveImportWhenTypeParametrizedByObjectInConstructor() = new FileSet {
+    """
+    package org.acme
+
+    trait Inner {
+      type Inner
+    }
+
+	  case class Toolbox[T <: Inner](inner: T)
```
Comment (mlangc) [inline src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala]: Is this (not correctly indented) case class necessary to trigger the problem?

File: src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala [hunk 8]
```diff
@@ -4410,4 +4411,902 @@ class OrganizeImportsTest extends OrganizeImportsBaseTest {
...
+  def shouldNotRemoveImportExtendedByWildcard_v3() = new FileSet {
+    """
+    package acme.duration
+
+	  trait DurationConvensions {
+      def baz = "baz"
+	  }
+
+	  object Duration {
+      def foo = "Duration"
```
Comment (mlangc) [inline src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala]: Wrong indentation

File: src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala (hunk 7 above)
Comment (wpopielarski) [inline src/test/scala/scala/tools/refactoring/tests/implementations/imports/OrganizeImportsTest.scala]: no, it is another failing scenario, sorry for confusion, should go with separate commit

**Comment by sschaef** [comment]
LGTM

**Pull request merged**